sql/src/main/java/org/apache/apex/malhar/sql/planner/RelNodeVisitor.java

```diff
@@ -0,0 +1,457 @@
+    } else if (n instanceof TableModify) {
+      currentNodeRelInfo = visitTableModify((TableModify)n, inputStreams);
+    } else if (n instanceof TableScan) {
+      currentNodeRelInfo = visitTableScan((TableScan)n, inputStreams);
+    } else if (n instanceof Uncollect) {
+      currentNodeRelInfo = visitUncollect((Uncollect)n, inputStreams);
+    } else if (n instanceof Window) {
+      currentNodeRelInfo = visitWindow((Window)n, inputStreams);
+    } else if (n instanceof Join) {
+      currentNodeRelInfo = visitJoin((Join)n, inputStreams);
+    } else {
+      currentNodeRelInfo = defaultValue(n, inputStreams);
+    }
+
+    if (currentNodeRelInfo != null && inputStreams.size() != 0) {
+      for (int i = 0; i < inputStreams.size(); i++) {
+        RelInfo inputStream = inputStreams.get(i);
+        Operator.OutputPort outputPort = inputStream.getOutPort();
+        Operator.InputPort inputPort = currentNodeRelInfo.getInputPorts().get(i);
+
+        String streamName = OperatorUtils.getUniqueStreamName(inputStream.getRelName(),
+            currentNodeRelInfo.getRelName());
+        Class schema;
+        if (inputStream.getOutRelDataType() != null) {
+          schema = getSchemaForRelDataType(streamName, inputStream.getOutRelDataType());
+        } else if (inputStream.getClazz() != null) {
+          schema = inputStream.getClazz();
+        } else {
+          throw new RuntimeException("Unexpected condition reached.");
+        }
+        dag.setOutputPortAttribute(outputPort, Context.PortContext.TUPLE_CLASS, schema);
+        dag.setInputPortAttribute(inputPort, Context.PortContext.TUPLE_CLASS, schema);
+        dag.addStream(streamName, outputPort, inputPort);
+      }
+    }
+
+    if (currentNodeRelInfo.getOutPort() == null) {
+      // End of the pipeline.
+      String schemaJar = tupleSchemaRegistry.generateCommonJar();
+
+      String jars = dag.getAttributes().get(Context.DAGContext.LIBRARY_JARS);
+      dag.setAttribute(Context.DAGContext.LIBRARY_JARS,
+          ((jars != null) && (jars.length() != 0)) ? jars + "," + schemaJar : schemaJar);
+    }
+
+    return currentNodeRelInfo;
+  }
+
+  public Class getSchemaForRelDataType(String schemaName, RelDataType rowType)
+  {
+    if (rowType.isStruct()) {
+      TupleSchemaRegistry.Schema newSchema = this.tupleSchemaRegistry.createNewSchema(schemaName);
+      for (RelDataTypeField field : rowType.getFieldList()) {
+        RelDataType type = field.getType();
+        newSchema.addField(OperatorUtils.getValidFieldName(field), convertPrimitiveToSqlType(type));
+      }
+      try {
+        newSchema.generateBean();
+      } catch (IOException | JSONException e) {
+        throw new RuntimeException("Failed t generate schema", e);
+      }
+      return newSchema.beanClass;
+    } else {
+      throw new RuntimeException("Yet to be implemented");
+    }
+  }
+
+  public RelInfo visitAggregate(Aggregate aggregate, List<RelInfo> inputStreams) throws Exception
```

Comment:
What's the purpose of that? I don't see that base class being used anywhere else. To me this looks appropriate as there are lesser numbers of files and lesser places to make changes on any further additions.